File: init-service/app/services/workspace_service.py
import os
import re
import shutil
import subprocess
import tempfile
import uuid
from datetime import UTC, datetime
import json

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session, joinedload

from app.db import session_scope
from app.models.project import Node, Project, Workspace
from app.services.cas_service import hash_file, upload_file_if_missing


INITIAL_INDEX_IGNORE_NAMES = {".git", ".venv", "venv", "__pycache__", "node_modules", ".pytest_cache"}
BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../.."))
WORKSPACES_DIR = os.path.join(BASE_DIR, "workspaces")
TEMPLATES_DIR = os.path.join(BASE_DIR, "runner", "templates")
TEMPLATE_MANIFESTS_DIR = os.path.join(BASE_DIR, "runner", "template_manifests")


def _now() -> datetime:
    return datetime.now(UTC)


def slugify(value: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", value.strip().lower()).strip("-")
    return slug or uuid.uuid4().hex[:8]


def workspace_disk_path(workspace_id: str) -> str:
    return os.path.join(WORKSPACES_DIR, workspace_id)


def project_disk_path(workspace_id: str, project_id: str) -> str:
    return os.path.join(workspace_disk_path(workspace_id), project_id)


def _serialize_workspace(workspace: Workspace) -> dict:
    return {
        "id": workspace.id,
        "ownerUserId": workspace.owner_user_id,
        "name": workspace.name,
        "slug": workspace.slug,
        "createdAt": workspace.created_at,
        "updatedAt": workspace.updated_at,
    }


def _serialize_project(project: Project) -> dict:
    return {
        "id": project.id,
        "workspaceId": project.workspace_id,
        "name": project.name,
        "slug": project.slug,
        "type": project.type,
        "createdAt": project.created_at,
        "updatedAt": project.updated_at,
    }


def _node_path(session: Session, node: Node) -> str:
    if node.parent_id is None:
        return "/"

    segments: list[str] = []
    current = node
    while current.parent_id is not None:
        segments.append(current.name)
        parent = session.get(Node, current.parent_id)
        if parent is None:
            break
        current = parent
    return "/" + "/".join(reversed(segments))


def serialize_node(session: Session, node: Node) -> dict:
    return {
        "id": node.id,
        "projectId": node.project_id,
        "parentId": node.parent_id,
        "name": node.name,
        "type": node.type,
        "path": _node_path(session, node),
        "contentHash": node.content_hash,
        "sizeBytes": node.size_bytes,
        "createdAt": node.created_at,
        "updatedAt": node.updated_at,
        "isRoot": node.parent_id is None,
    }


def _validate_name(name: str) -> str:
    normalized = name.strip()
    if not normalized:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Name cannot be empty.")
    if "/" in normalized or "\\" in normalized:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Name cannot contain path separators.")
    return normalized


def _content_hash_for_file(abs_path: str) -> tuple[str | None, int | None]:
    try:
        content_hash, size_bytes = hash_file(abs_path)
    except OSError:
        return None, None
    upload_file_if_missing(content_hash, abs_path)
    return content_hash, size_bytes


def _create_root_node(session: Session, project: Project) -> Node:
    now = _now()
    root_node = Node(
        id=str(uuid.uuid4()),
        project_id=project.id,
        parent_id=None,
        name=project.slug,
        type="FOLDER",
        content_hash=None,
        size_bytes=None,
        created_at=now,
        updated_at=now,
    )
    session.add(root_node)
    session.flush()
    return root_node


def _index_tree(session: Session, project: Project, root_node: Node, project_dir: str) -> None:
    folder_ids: dict[str, str] = {"": root_node.id}
    now = _now()

    for current_root, dir_names, file_names in os.walk(project_dir):
        dir_names[:] = [name for name in dir_names if name not in INITIAL_INDEX_IGNORE_NAMES]
        rel_dir = os.path.relpath(current_root, project_dir)
        rel_dir = "" if rel_dir == "." else rel_dir.replace(os.sep, "/")

        if rel_dir:
            parent_rel = os.path.dirname(rel_dir).replace(os.sep, "/")
            node = Node(
                id=str(uuid.uuid4()),
                project_id=project.id,
                parent_id=folder_ids.get(parent_rel, root_node.id),
                name=os.path.basename(current_root),
                type="FOLDER",
                content_hash=None,
                size_bytes=None,
                created_at=now,
                updated_at=now,
            )
            session.add(node)
            session.flush()
            folder_ids[rel_dir] = node.id

        for file_name in file_names:
            if file_name in INITIAL_INDEX_IGNORE_NAMES:
                continue
            abs_file = os.path.join(current_root, file_name)
            rel_file = os.path.relpath(abs_file, project_dir).replace(os.sep, "/")
            parent_rel = os.path.dirname(rel_file).replace(os.sep, "/")
            content_hash, size_bytes = _content_hash_for_file(abs_file)
            session.add(
                Node(
                    id=str(uuid.uuid4()),
                    project_id=project.id,
                    parent_id=folder_ids.get(parent_rel, root_node.id) if parent_rel else root_node.id,
                    name=file_name,
                    type="FILE",
                    content_hash=content_hash,
                    size_bytes=size_bytes,
                    created_at=now,
                    updated_at=now,
                )
            )


def _manifest_path(project_type: str) -> str:
    return os.path.join(TEMPLATE_MANIFESTS_DIR, f"{project_type}.json")


def _load_template_manifest(project_type: str) -> dict | None:
    manifest_path = _manifest_path(project_type)
    if not os.path.exists(manifest_path):
        return None
    try:
        with open(manifest_path, "r", encoding="utf-8") as handle:
            manifest = json.load(handle)
    except (OSError, json.JSONDecodeError):
        return None
    if manifest.get("version") != 1 or manifest.get("templateType") != project_type:
        return None
    if not isinstance(manifest.get("directories"), list) or not isinstance(manifest.get("files"), list):
        return None
    return manifest


def _index_tree_from_manifest(session: Session, project: Project, root_node: Node, manifest: dict) -> None:
    folder_ids: dict[str, str] = {"": root_node.id}
    now = _now()

    directories = sorted(
        (
            entry.get("path", "").strip("/")
            for entry in manifest.get("directories", [])
            if isinstance(entry, dict)
        ),
        key=lambda path: (path.count("/"), path),
    )
    for rel_dir in directories:
        if not rel_dir:
            continue
        parent_rel = os.path.dirname(rel_dir).replace(os.sep, "/")
        node = Node(
            id=str(uuid.uuid4()),
            project_id=project.id,
            parent_id=folder_ids.get(parent_rel, root_node.id),
            name=os.path.basename(rel_dir),
            type="FOLDER",
            content_hash=None,
            size_bytes=None,
            created_at=now,
            updated_at=now,
        )
        session.add(node)
        session.flush()
        folder_ids[rel_dir] = node.id

    for entry in manifest.get("files", []):
        if not isinstance(entry, dict):
            continue
        rel_file = str(entry.get("path", "")).strip("/")
        if not rel_file:
            continue
        parent_rel = os.path.dirname(rel_file).replace(os.sep, "/")
        session.add(
            Node(
                id=str(uuid.uuid4()),
                project_id=project.id,
                parent_id=folder_ids.get(parent_rel, root_node.id) if parent_rel else root_node.id,
                name=os.path.basename(rel_file),
                type="FILE",
                content_hash=str(entry.get("contentHash") or "") or None,
                size_bytes=entry.get("sizeBytes") if isinstance(entry.get("sizeBytes"), int) else None,
                created_at=now,
                updated_at=now,
            )
        )


def _copy_template(project_type: str, target_dir: str) -> None:
    template_dir = os.path.join(TEMPLATES_DIR, project_type)
    if not os.path.exists(template_dir):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Unsupported project type '{project_type}'.",
        )
    shutil.copytree(
        template_dir,
        target_dir,
        dirs_exist_ok=True,
        ignore=shutil.ignore_patterns(*INITIAL_INDEX_IGNORE_NAMES),
    )


def _clone_repository(github_url: str, target_dir: str) -> None:
    if not (
        github_url.startswith("https://github.com/")
        or github_url.startswith("http://github.com/")
        or github_url.startswith("git@github.com:")
    ):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Only GitHub repositories are supported.",
        )

    tmp_dir = tempfile.mkdtemp(prefix="yuvro_clone_")
    clone_target = os.path.join(tmp_dir, "repo")
    try:
        result = subprocess.run(
            ["git", "clone", "--depth", "1", github_url, clone_target],
            capture_output=True,
            text=True,
            timeout=120,
        )
        if result.returncode != 0:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=result.stderr.strip() or result.stdout.strip() or "git clone failed.",
            )
        shutil.copytree(clone_target, target_dir, dirs_exist_ok=True, ignore=shutil.ignore_patterns(".git"))
    finally:
        shutil.rmtree(tmp_dir, ignore_errors=True)


def _bootstrap_workspace_and_project(
    *,
    owner_user_id: str,
    workspace_name: str,
    project_name: str,
    project_type: str,
    bootstrap_fn,
) -> dict:
    workspace_id = str(uuid.uuid4())
    project_id = str(uuid.uuid4())
    now = _now()

    workspace = Workspace(
        id=workspace_id,
        owner_user_id=owner_user_id,
        name=workspace_name.strip(),
        slug=slugify(workspace_name),
        created_at=now,
        updated_at=now,
    )
    project = Project(
        id=project_id,
        workspace_id=workspace_id,
        name=project_name.strip(),
        slug=slugify(project_name),
        type=project_type,
        created_at=now,
        updated_at=now,
    )

    project_dir = project_disk_path(workspace_id, project_id)
    os.makedirs(project_dir, exist_ok=True)
    bootstrap_fn(project_dir)

    with session_scope() as session:
        session.add(workspace)
        session.add(project)
        session.flush()
        root_node = _create_root_node(session, project)
        manifest = _load_template_manifest(project_type) if project_type != "github" else None
        if manifest:
            _index_tree_from_manifest(session, project, root_node, manifest)
        else:
            _index_tree(session, project, root_node, project_dir)
        session.flush()
        return {
            "workspace": _serialize_workspace(workspace),
            "project": _serialize_project(project),
            "rootNode": serialize_node(session, root_node),
        }


def create_template_project(owner_user_id: str, workspace_name: str, project_name: str, project_type: str) -> dict:
    return _bootstrap_workspace_and_project(
        owner_user_id=owner_user_id,
        workspace_name=workspace_name,
        project_name=project_name,
        project_type=project_type,
        bootstrap_fn=lambda project_dir: _copy_template(project_type, project_dir),
    )


def clone_project(owner_user_id: str, workspace_name: str, project_name: str, github_url: str) -> dict:
    return _bootstrap_workspace_and_project(
        owner_user_id=owner_user_id,
        workspace_name=workspace_name,
        project_name=project_name,
        project_type="github",
        bootstrap_fn=lambda project_dir: _clone_repository(github_url, project_dir),
    )


def get_project_detail(owner_user_id: str, project_id: str) -> dict:
    with session_scope() as session:
        project = session.scalar(
            select(Project)
            .options(joinedload(Project.workspace))
            .where(Project.id == project_id)
        )
        if project is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found.")
        if project.workspace.owner_user_id != owner_user_id:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found.")
        root_node = session.scalar(
            select(Node).where(Node.project_id == project.id, Node.parent_id.is_(None))
        )
        if root_node is None:
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Project root node is missing.")
        return {
            "workspace": _serialize_workspace(project.workspace),
            "project": _serialize_project(project),
            "rootNode": serialize_node(session, root_node),
        }


def get_project_nodes(owner_user_id: str, project_id: str) -> dict:
    with session_scope() as session:
        project = session.scalar(
            select(Project)
            .options(joinedload(Project.workspace))
            .where(Project.id == project_id)
        )
        if project is None or project.workspace.owner_user_id != owner_user_id:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found.")
        nodes = session.scalars(
            select(Node).where(Node.project_id == project_id).order_by(Node.parent_id.is_not(None), Node.type.desc(), Node.name)
        ).all()
        return {"nodes": [serialize_node(session, node) for node in nodes]}


def get_node(session: Session, node_id: str, owner_user_id: str | None = None) -> Node:
    node = session.scalar(
        select(Node)
        .options(joinedload(Node.project).joinedload(Project.workspace))
        .where(Node.id == node_id)
    )
    if node is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Node not found.")
    if owner_user_id and node.project.workspace.owner_user_id != owner_user_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Node not found.")
    return node


def relative_path_for_node(session: Session, node: Node) -> str:
    return _node_path(session, node).lstrip("/")

File: init-service/src/api/project/workspace-service.ts
import path from "path";
import fs from "fs";
import os from "os";
import { exec } from "child_process";
import { promisify } from "util";
import { eq, and, desc, isNull } from "drizzle-orm";
import { db } from "../../loaders/postgres";
import { workspaces as workspacesTable, projects as projectsTable, nodes as nodesTable } from "db-schema";
import config from "../../config";
import ApiError from "../../utils/ApiError";
import { hashFile, uploadFileIfMissing } from "../../shared/cas-service";
import { fileURLToPath } from "url";

const execAsync = promisify(exec);
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const INITIAL_INDEX_IGNORE_NAMES = new Set([".git", ".venv", "venv", "__pycache__", "node_modules", ".pytest_cache"]);
const BASE_DIR = path.resolve(__dirname, "../../../..");
export const WORKSPACES_DIR = process.env.WORKSPACES_DIR || path.join(BASE_DIR, "workspaces");
export const TEMPLATES_DIR = process.env.TEMPLATES_DIR || path.join(BASE_DIR, "runner", "templates");
export const TEMPLATE_MANIFESTS_DIR = process.env.TEMPLATE_MANIFESTS_DIR || path.join(BASE_DIR, "runner", "template_manifests");

export function slugify(value: string): string {
  const slug = value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
  return slug || crypto.randomUUID().substring(0, 8);
}

export function workspaceDiskPath(workspaceId: string): string {
  return path.join(WORKSPACES_DIR, workspaceId);
}

export function projectDiskPath(workspaceId: string, projectId: string): string {
  return path.join(workspaceDiskPath(workspaceId), projectId);
}

function serializeWorkspace(workspace: any): any {
  return {
    id: workspace.id,
    ownerUserId: workspace.ownerUserId,
    name: workspace.name,
    slug: workspace.slug,
    createdAt: workspace.createdAt.toISOString(),
    updatedAt: workspace.updatedAt.toISOString(),
  };
}

function serializeProject(project: any): any {
  return {
    id: project.id,
    workspaceId: project.workspaceId,
    name: project.name,
    slug: project.slug,
    type: project.type,
    createdAt: project.createdAt.toISOString(),
    updatedAt: project.updatedAt.toISOString(),
  };
}

function serializeWorkspaceWithProjects(workspace: any, projects: any[]): any {
  const sortedProjects = [...projects].sort((a, b) => {
    const timeDiff = a.createdAt.getTime() - b.createdAt.getTime();
    if (timeDiff !== 0) return timeDiff;
    return a.name.toLowerCase().localeCompare(b.name.toLowerCase());
  });

  return {
    ...serializeWorkspace(workspace),
    projects: sortedProjects.map(serializeProject),
  };
}

// Calculates segment-based node path e.g. "/src/main.js"
export async function nodePath(node: any, tx: any = db): Promise<string> {
  if (!node.parentId) {
    return "/";
  }

  const segments: string[] = [];
  let current = node;

  while (current.parentId) {
    segments.push(current.name);
    const parents = await tx
      .select()
      .from(nodesTable)
      .where(eq(nodesTable.id, current.parentId));

    const parent = parents[0];
    if (!parent) break;
    current = parent;
  }

  return "/" + segments.reverse().join("/");
}

export async function serializeNode(node: any, tx: any = db): Promise<any> {
  const pathVal = await nodePath(node, tx);
  return {
    id: node.id,
    projectId: node.projectId,
    parentId: node.parentId,
    name: node.name,
    type: node.type,
    path: pathVal,
    contentHash: node.contentHash,
    sizeBytes: node.sizeBytes,
    createdAt: node.createdAt.toISOString(),
    updatedAt: node.updatedAt.toISOString(),
    isRoot: !node.parentId,
  };
}

export function validateName(name: string): string {
  const normalized = name.trim();
  if (!normalized) {
    throw new ApiError("Name cannot be empty.", 422);
  }
  if (normalized.includes("/") || normalized.includes("\\")) {
    throw new ApiError("Name cannot contain path separators.", 422);
  }
  return normalized;
}

export async function getWorkspaceForOwner(ownerUserId: string, workspaceId: string, tx: any = db): Promise<any> {
  const workspaces = await tx
    .select()
    .from(workspacesTable)
    .where(
      and(
        eq(workspacesTable.id, workspaceId),
        eq(workspacesTable.ownerUserId, ownerUserId)
      )
    );

  const workspace = workspaces[0];
  if (!workspace) {
    throw new ApiError("Workspace not found.", 404);
  }
  return workspace;
}

async function contentHashForFile(absPath: string): Promise<[string | null, number | null]> {
  try {
    const [contentHash, sizeBytes] = await hashFile(absPath);
    await uploadFileIfMissing(contentHash, absPath);
    return [contentHash, sizeBytes];
  } catch {
    return [null, null];
  }
}

async function createRootNode(tx: any, project: any): Promise<any> {
  const now = new Date();
  const rootNode = {
    id: crypto.randomUUID(),
    projectId: project.id,
    parentId: null,
    name: project.slug,
    type: "FOLDER",
    contentHash: null,
    sizeBytes: null,
    createdAt: now,
    updatedAt: now,
  };

  await tx.insert(nodesTable).values(rootNode);
  return rootNode;
}

// Recursively walks through projectDir and populates the database Node representations
async function indexTree(tx: any, project: any, rootNode: any, projectDir: string): Promise<void> {
  const folderIds: Record<string, string> = { "": rootNode.id };
  const now = new Date();

  function walk(currentDir: string) {
    const list = fs.readdirSync(currentDir);
    const relDirRaw = path.relative(projectDir, currentDir);
    const relDir = relDirRaw === "" ? "" : relDirRaw.replace(/\\/g, "/");

    const dirs: string[] = [];
    const files: string[] = [];

    for (const file of list) {
      if (INITIAL_INDEX_IGNORE_NAMES.has(file)) continue;
      const fullPath = path.join(currentDir, file);
      const stat = fs.statSync(fullPath);
      if (stat.isDirectory()) {
        dirs.push(file);
      } else {
        files.push(file);
      }
    }

    return { relDir, dirs, files, currentDir };
  }

  // Use a queue to implement BFS/DFS tree scan
  const queue = [projectDir];
  while (queue.length > 0) {
    const current = queue.shift()!;
    const { relDir, dirs, files } = walk(current);

    if (relDir) {
      const parentRel = path.dirname(relDir).replace(/\\/g, "/");
      const parentRelNormalized = parentRel === "." ? "" : parentRel;

      const nodeId = crypto.randomUUID();
      const node = {
        id: nodeId,
        projectId: project.id,
        parentId: folderIds[parentRelNormalized] || rootNode.id,
        name: path.basename(current),
        type: "FOLDER",
        contentHash: null,
        sizeBytes: null,
        createdAt: now,
        updatedAt: now,
      };
      await tx.insert(nodesTable).values(node);
      folderIds[relDir] = nodeId;
    }

    for (const file of files) {
      const absFile = path.join(current, file);
      const relFile = path.relative(projectDir, absFile).replace(/\\/g, "/");
      const parentRel = path.dirname(relFile).replace(/\\/g, "/");
      const parentRelNormalized = parentRel === "." ? "" : parentRel;

      const [contentHash, sizeBytes] = await contentHashForFile(absFile);
      await tx.insert(nodesTable).values({
        id: crypto.randomUUID(),
        projectId: project.id,
        parentId: folderIds[parentRelNormalized] || rootNode.id,
        name: file,
        type: "FILE",
        contentHash,
        sizeBytes,
        createdAt: now,
        updatedAt: now,
      });
    }

    for (const dir of dirs) {
      queue.push(path.join(current, dir));
    }
  }
}

function loadTemplateManifest(projectType: string): any | null {
  const manifestPath = path.join(TEMPLATE_MANIFESTS_DIR, `${projectType}.json`);
  if (!fs.existsSync(manifestPath)) return null;

  try {
    const data = JSON.parse(fs.readFileSync(manifestPath, "utf-8"));
    if (data.version !== 1 || data.templateType !== projectType) return null;
    if (!Array.isArray(data.directories) || !Array.isArray(data.files)) return null;
    return data;
  } catch {
    return null;
  }
}

async function indexTreeFromManifest(tx: any, project: any, rootNode: any, manifest: any): Promise<void> {
  const folderIds: Record<string, string> = { "": rootNode.id };
  const now = new Date();

  // Sort directories by path segment count to create parents before children
  const directories = (manifest.directories || [])
    .map((entry: any) => (entry.path || "").replace(/^\/|\/$/g, ""))
    .filter(Boolean)
    .sort((a: string, b: string) => {
      const aCount = (a.match(/\//g) || []).length;
      const bCount = (b.match(/\//g) || []).length;
      if (aCount !== bCount) return aCount - bCount;
      return a.localeCompare(b);
    });

  for (const relDir of directories) {
    const parentRel = path.dirname(relDir).replace(/\\/g, "/");
    const parentRelNormalized = parentRel === "." ? "" : parentRel;

    const nodeId = crypto.randomUUID();
    await tx.insert(nodesTable).values({
      id: nodeId,
      projectId: project.id,
      parentId: folderIds[parentRelNormalized] || rootNode.id,
      name: path.basename(relDir),
      type: "FOLDER",
      contentHash: null,
      sizeBytes: null,
      createdAt: now,
      updatedAt: now,
    });
    folderIds[relDir] = nodeId;
  }

  for (const entry of manifest.files || []) {
    const relFile = (entry.path || "").replace(/^\/|\/$/g, "");
    if (!relFile) continue;

    const parentRel = path.dirname(relFile).replace(/\\/g, "/");
    const parentRelNormalized = parentRel === "." ? "" : parentRel;

    await tx.insert(nodesTable).values({
      id: crypto.randomUUID(),
      projectId: project.id,
      parentId: folderIds[parentRelNormalized] || rootNode.id,
      name: path.basename(relFile),
      type: "FILE",
      contentHash: entry.contentHash || null,
      sizeBytes: typeof entry.sizeBytes === "number" ? entry.sizeBytes : null,
      createdAt: now,
      updatedAt: now,
    });
  }
}

function copyTemplate(projectType: string, targetDir: string): void {
  const templateDir = path.join(TEMPLATES_DIR, projectType);
  if (!fs.existsSync(templateDir)) {
    throw new ApiError(`Unsupported project type '${projectType}'.`, 422);
  }

  fs.mkdirSync(targetDir, { recursive: true });
  fs.cpSync(templateDir, targetDir, {
    recursive: true,
    filter: (src) => {
      const name = path.basename(src);
      return !INITIAL_INDEX_IGNORE_NAMES.has(name);
    },
  });
}

async function cloneRepository(githubUrl: string, targetDir: string): Promise<void> {
  if (
    !githubUrl.startsWith("https://github.com/") &&
    !githubUrl.startsWith("http://github.com/") &&
    !githubUrl.startsWith("git@github.com:")
  ) {
    throw new ApiError("Only GitHub repositories are supported.", 422);
  }

  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "yuvro_clone_"));
  const cloneTarget = path.join(tmpDir, "repo");

  try {
    await execAsync(`git clone --depth 1 "${githubUrl}" "${cloneTarget}"`);
    fs.mkdirSync(targetDir, { recursive: true });

    // Copy excluding .git
    fs.cpSync(cloneTarget, targetDir, {
      recursive: true,
      filter: (src) => {
        const name = path.basename(src);
        return name !== ".git";
      },
    });
  } catch (error: any) {
    throw new ApiError(error.stderr?.trim() || "git clone failed.", 400);
  } finally {
    try {
      fs.rmSync(tmpDir, { recursive: true, force: true });
    } catch {}
  }
}

async function bootstrapProject(
  workspace: any,
  projectName: string,
  projectType: string,
  bootstrapFn: (projectDir: string) => Promise<void> | void
): Promise<any> {
  const now = new Date();
  const normalizedName = validateName(projectName);
  const projectId = crypto.randomUUID();

  const project = {
    id: projectId,
    workspaceId: workspace.id,
    name: normalizedName,
    slug: slugify(normalizedName),
    type: projectType,
    createdAt: now,
    updatedAt: now,
  };

  const projectDir = projectDiskPath(workspace.id, projectId);
  fs.mkdirSync(workspaceDiskPath(workspace.id), { recursive: true });
  fs.mkdirSync(projectDir, { recursive: true });

  try {
    await bootstrapFn(projectDir);

    return await db.transaction(async (tx) => {
      // Fetch and update workspace updatedAt
      await tx
        .update(workspacesTable)
        .set({ updatedAt: now })
        .where(eq(workspacesTable.id, workspace.id));

      await tx.insert(projectsTable).values(project);

      const rootNode = await createRootNode(tx, project);
      const manifest = projectType !== "github" ? loadTemplateManifest(projectType) : null;

      if (manifest) {
        await indexTreeFromManifest(tx, project, rootNode, manifest);
      } else {
        await indexTree(tx, project, rootNode, projectDir);
      }

      const serializedNode = await serializeNode(rootNode, tx);

      return {
        workspace: serializeWorkspace(workspace),
        project: serializeProject(project),
        rootNode: serializedNode,
      };
    });
  } catch (error: any) {
    try {
      fs.rmSync(projectDir, { recursive: true, force: true });
    } catch {}

    // Check unique constraint conflict
    if (error.code === "23505" || error.message?.includes("uq_project_workspace_name")) {
      throw new ApiError(`A project named '${project.name}' already exists in this workspace.`, 409);
    }
    throw error;
  }
}

async function bootstrapWorkspaceAndProject(
  ownerUserId: string,
  workspaceName: string,
  projectName: string,
  projectType: string,
  bootstrapFn: (projectDir: string) => Promise<void> | void
): Promise<any> {
  const now = new Date();
  const normalizedWorkspaceName = validateName(workspaceName);
  const workspaceId = crypto.randomUUID();

  const workspace = {
    id: workspaceId,
    ownerUserId,
    name: normalizedWorkspaceName,
    slug: slugify(normalizedWorkspaceName),
    createdAt: now,
    updatedAt: now,
  };

  fs.mkdirSync(workspaceDiskPath(workspaceId), { recursive: true });

  try {
    await db.insert(workspacesTable).values(workspace);
    return await bootstrapProject(workspace, projectName, projectType, bootstrapFn);
  } catch (error) {
    // Cleanup workspace
    await db.delete(workspacesTable).where(eq(workspacesTable.id, workspaceId));
    try {
      fs.rmSync(workspaceDiskPath(workspaceId), { recursive: true, force: true });
    } catch {}
    throw error;
  }
}

// ─── Exported API Functions ──────────────────────────────────────────────────

export async function createTemplateProject(
  ownerUserId: string,
  workspaceName: string,
  projectName: string,
  projectType: string
): Promise<any> {
  return bootstrapWorkspaceAndProject(
    ownerUserId,
    workspaceName,
    projectName,
    projectType,
    (projectDir) => copyTemplate(projectType, projectDir)
  );
}

export async function cloneProject(
  ownerUserId: string,
  workspaceName: string,
  projectName: string,
  githubUrl: string
): Promise<any> {
  return bootstrapWorkspaceAndProject(
    ownerUserId,
    workspaceName,
    projectName,
    "github",
    (projectDir) => cloneRepository(githubUrl, projectDir)
  );
}

export async function createTemplateProjectInWorkspace(
  ownerUserId: string,
  workspaceId: string,
  projectName: string,
  projectType: string
): Promise<any> {
  const workspace = await getWorkspaceForOwner(ownerUserId, workspaceId);
  return bootstrapProject(workspace, projectName, projectType, (projectDir) =>
    copyTemplate(projectType, projectDir)
  );
}

export async function cloneProjectInWorkspace(
  ownerUserId: string,
  workspaceId: string,
  projectName: string,
  githubUrl: string
): Promise<any> {
  const workspace = await getWorkspaceForOwner(ownerUserId, workspaceId);
  return bootstrapProject(workspace, projectName, "github", (projectDir) =>
    cloneRepository(githubUrl, projectDir)
  );
}

export async function listWorkspaces(ownerUserId: string): Promise<any> {
  const workspaces = await db
    .select()
    .from(workspacesTable)
    .where(eq(workspacesTable.ownerUserId, ownerUserId))
    .orderBy(desc(workspacesTable.updatedAt), desc(workspacesTable.createdAt));

  const workspacesWithProjects = await Promise.all(
    workspaces.map(async (workspace) => {
      const projects = await db
        .select()
        .from(projectsTable)
        .where(eq(projectsTable.workspaceId, workspace.id));
      return serializeWorkspaceWithProjects(workspace, projects);
    })
  );

  return { workspaces: workspacesWithProjects };
}

export async function getProjectDetail(ownerUserId: string, projectId: string): Promise<any> {
  const projects = await db
    .select({
      id: projectsTable.id,
      workspaceId: projectsTable.workspaceId,
      name: projectsTable.name,
      slug: projectsTable.slug,
      type: projectsTable.type,
      createdAt: projectsTable.createdAt,
      updatedAt: projectsTable.updatedAt,
      workspace: {
        id: workspacesTable.id,
        ownerUserId: workspacesTable.ownerUserId,
        name: workspacesTable.name,
        slug: workspacesTable.slug,
        createdAt: workspacesTable.createdAt,
        updatedAt: workspacesTable.updatedAt,
      },
    })
    .from(projectsTable)
    .innerJoin(workspacesTable, eq(projectsTable.workspaceId, workspacesTable.id))
    .where(eq(projectsTable.id, projectId));

  const project = projects[0];
  if (!project || project.workspace.ownerUserId !== ownerUserId) {
    throw new ApiError("Project not found.", 404);
  }

  const rootNodes = await db
    .select()
    .from(nodesTable)
    .where(and(eq(nodesTable.projectId, projectId), isNull(nodesTable.parentId)));

  const rootNode = rootNodes[0];
  if (!rootNode) {
    throw new ApiError("Project root node is missing.", 500);
  }

  return {
    workspace: serializeWorkspace(project.workspace),
    project: serializeProject(project),
    rootNode: await serializeNode(rootNode),
  };
}

export async function getProjectNodes(ownerUserId: string, projectId: string): Promise<any> {
  const projects = await db
    .select({
      id: projectsTable.id,
      workspace: {
        ownerUserId: workspacesTable.ownerUserId,
      },
    })
    .from(projectsTable)
    .innerJoin(workspacesTable, eq(projectsTable.workspaceId, workspacesTable.id))
    .where(eq(projectsTable.id, projectId));

  const project = projects[0];
  if (!project || project.workspace.ownerUserId !== ownerUserId) {
    throw new ApiError("Project not found.", 404);
  }

  const nodes = await db
    .select()
    .from(nodesTable)
    .where(eq(nodesTable.projectId, projectId))
    .orderBy(desc(nodesTable.parentId), desc(nodesTable.type), nodesTable.name); // folder before files

  const serializedNodes = await Promise.all(nodes.map((n) => serializeNode(n)));
  return { nodes: serializedNodes };
}

export async function getNode(nodeId: string, ownerUserId?: string, tx: any = db): Promise<any> {
  const nodes = await tx
    .select({
      id: nodesTable.id,
      projectId: nodesTable.projectId,
      parentId: nodesTable.parentId,
      name: nodesTable.name,
      type: nodesTable.type,
      contentHash: nodesTable.contentHash,
      sizeBytes: nodesTable.sizeBytes,
      createdAt: nodesTable.createdAt,
      updatedAt: nodesTable.updatedAt,
      project: {
        id: projectsTable.id,
        workspaceId: projectsTable.workspaceId,
        workspace: {
          id: workspacesTable.id,
          ownerUserId: workspacesTable.ownerUserId,
        },
      },
    })
    .from(nodesTable)
    .innerJoin(projectsTable, eq(nodesTable.projectId, projectsTable.id))
    .innerJoin(workspacesTable, eq(projectsTable.workspaceId, workspacesTable.id))
    .where(eq(nodesTable.id, nodeId));

  const node = nodes[0];
  if (!node) {
    throw new ApiError("Node not found.", 404);
  }

  if (ownerUserId && node.project.workspace.ownerUserId !== ownerUserId) {
    throw new ApiError("Node not found.", 404);
  }

  return node;
}

export async function relativePathForNode(node: any, tx: any = db): Promise<string> {
  const pathVal = await nodePath(node, tx);
  return pathVal.replace(/^\//, "");
}
